Implement INSERT for Hive with partition support

dain opened the pull request.

dain commented:
For anyone with time, please review and test this out in your environment.

yuananf commented:
There are tests failures, ```TestTpchDistributedQueriesWithoutHashGeneration``` should also use ```testSessionBuilder()``` instead of ```Session.builder()```

dain commented:
Should be fixed now.  Someone checked in some new files and I missed them in the rebase.

nezihyigitbasi commented:
@dain Just hit a classpath issue while running some tests. I guess the reason is a main module class is trying to load classes from the hive plugin, hence failing.

```
presto:nyigitbasi> insert into clickmodel select * from clickmodel where countrygroup='C2';

Query 20150730_045412_00028_29kez, FAILED, 1 node
http://localhost:8080/v1/query/20150730_045412_00028_29kez?pretty
Splits: 4 total, 0 done (0.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 33% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
0:03 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20150730_045412_00028_29kez failed: Unable to create input format org.apache.hadoop.mapred.TextInputFormat
java.lang.RuntimeException: Unable to create input format org.apache.hadoop.mapred.TextInputFormat
	at com.facebook.presto.hive.HiveUtil.getInputFormat(HiveUtil.java:188)
	at com.facebook.presto.hive.HiveUtil.createRecordReader(HiveUtil.java:142)
	at com.facebook.presto.hive.GenericHiveRecordCursorProvider.createHiveRecordCursor(GenericHiveRecordCursorProvider.java:47)
	at com.facebook.presto.hive.HivePageSourceProvider.getHiveRecordCursor(HivePageSourceProvider.java:128)
	at com.facebook.presto.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:106)
	at com.facebook.presto.spi.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:43)
	at com.facebook.presto.split.PageSourceManager.createPageSource(PageSourceManager.java:48)
	at com.facebook.presto.operator.TableScanOperator.createSourceIfNecessary(TableScanOperator.java:258)
	at com.facebook.presto.operator.TableScanOperator.isFinished(TableScanOperator.java:206)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:377)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:303)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:587)
	at com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:505)
	at com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:639)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.mapred.TextInputFormat not found
	at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:1626)
	at com.facebook.presto.hive.HiveUtil.getInputFormatClass(HiveUtil.java:202)
	at com.facebook.presto.hive.HiveUtil.getInputFormat(HiveUtil.java:179)
	... 16 more
```

dain commented:
@nezihyigitbasi that exception is coming from the read path, and I don't think I touched that.  

The really weird thing is the code is passing though the `ClassLoaderSafeConnectorPageSourceProvider` which sets the thread context class loader.

nezihyigitbasi commented:
@dain The ugly thing is when I restart presto & re-run the exact same query I got another exception, seems like this time it's on the write path.

```
presto:nyigitbasi>  insert into clickmodel select * from clickmodel where countrygroup='C2';

Query 20150730_051734_00005_44c56, FAILED, 1 node
http://localhost:8080/v1/query/20150730_051734_00005_44c56?pretty
Splits: 4 total, 2 done (50.00%)
CPU Time: 0.1s total,    75 rows/s, 2.56KB/s, 39% active
Per Node: 0.0 parallelism,     3 rows/s,   130B/s
Parallelism: 0.0
0:03 [10 rows, 346B] [3 rows/s, 130B/s]

Query 20150730_051734_00005_44c56 failed: null
java.lang.ExceptionInInitializerError
	at org.apache.hadoop.hive.shims.ShimLoader.getMajorVersion(ShimLoader.java:160)
	at org.apache.hadoop.hive.shims.ShimLoader.loadShims(ShimLoader.java:139)
	at org.apache.hadoop.hive.shims.ShimLoader.getHadoopShims(ShimLoader.java:100)
	at org.apache.hadoop.hive.conf.HiveConf$ConfVars.<clinit>(HiveConf.java:368)
	at org.apache.hadoop.hive.serde2.lazy.LazySerDeParameters.<init>(LazySerDeParameters.java:103)
	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.initialize(LazySimpleSerDe.java:125)
	at com.facebook.presto.hive.HivePageSink$HiveRecordWriter.initializeSerializer(HivePageSink.java:580)
	at com.facebook.presto.hive.HivePageSink$HiveRecordWriter.<init>(HivePageSink.java:517)
	at com.facebook.presto.hive.HivePageSink.createWriter(HivePageSink.java:394)
	at com.facebook.presto.hive.HivePageSink.appendPage(HivePageSink.java:260)
	at com.facebook.presto.operator.TableWriterOperator.addInput(TableWriterOperator.java:176)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:386)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:303)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:587)
	at com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:505)
	at com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:639)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at java.util.Properties$LineReader.readLine(Properties.java:434)
	at java.util.Properties.load0(Properties.java:353)
	at java.util.Properties.load(Properties.java:341)
	at org.apache.hadoop.util.VersionInfo.<init>(VersionInfo.java:51)
	at org.apache.hadoop.util.VersionInfo.<clinit>(VersionInfo.java:97)
	... 19 more
```

dain commented:
That stack is really in the initialization of the HiveConf$ConfVars and then VersionInfo.  Looks like the code is looking for a file "common-version-info.properties" in the thread context class loader.  

The problem is we don't have a `ClassLoaderSafeConnectorPageSink` yet

yuananf commented:
We met this problem before, and we solved this by adding code as below:
```
try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
return getRecordWriter(path);
}, 
```

yuananf commented:
When I run a sql below
```
create table employees_insert as select * from employees
```
It throws error

```
java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.hdfs.DistributedFileSystem not found
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:1720)
	at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2415)
	at org.apache.hadoop.fs.PrestoFileSystemCache.createFileSystem(PrestoFileSystemCache.java:69)
	at org.apache.hadoop.fs.PrestoFileSystemCache.getInternal(PrestoFileSystemCache.java:61)
	at org.apache.hadoop.fs.PrestoFileSystemCache.get(PrestoFileSystemCache.java:43)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:367)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:287)
	at com.facebook.presto.hive.HdfsEnvironment.getFileSystem(HdfsEnvironment.java:53)
	at com.facebook.presto.hive.HiveWriteUtils.pathExists(HiveWriteUtils.java:328)
	at com.facebook.presto.hive.HiveWriteUtils.getTableDefaultLocation(HiveWriteUtils.java:310)
	at com.facebook.presto.hive.HivePageSink.createWriter(HivePageSink.java:296)
	at com.facebook.presto.hive.HivePageSink.appendPage(HivePageSink.java:260)
	at com.facebook.presto.operator.TableWriterOperator.addInput(TableWriterOperator.java:176)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:386)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:303)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:587)
	at com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:505)
	at com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:639)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.hdfs.DistributedFileSystem not found
	at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:1626)
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:1718)
	... 20 more
```

dain commented:
I think I have fixed the class loader issue and updated this PR.  I haven't finished the testing yet, so it might still be broken.  I expect to finish my testing tomorrow sometime.

yuananf commented:
Works fine now, very good!

sogorkis commented:
I have tested this patch on eight node cluster and encountered problem when there is more than one `INSERT INTO ... SELECT` statement happening at the same time.

Way to reproduce this problem:

- Create two TPCH `lineitem` like tables:
```
CREATE TABLE hive.default.lineitem_insert_test_1 WITH (format='ORC') AS SELECT * FROM "tpch"."sf10"."lineitem" LIMIT 0;
CREATE TABLE hive.default.lineitem_insert_test_2 WITH (format='ORC') AS SELECT * FROM "tpch"."sf10"."lineitem" LIMIT 0;
```

- Run two inserts in parallel. I am using as a source lineitem table from 100gb TPCH schema which is in ORC format. Selecting 10% of rows:
```
INSERT INTO hive.default.lineitem_insert_test_1 SELECT * FROM "hive"."tpch_100gb_orc"."lineitem" WHERE quantity < 6;
INSERT INTO hive.default.lineitem_insert_test_2 SELECT * FROM "hive"."tpch_100gb_orc"."lineitem" WHERE quantity < 6;
```

- Queries fail with `INTERNAL_ERROR - PAGE_TRANSPORT_TIMEOUT (65540)`. Stacktrace:
```
com.facebook.presto.operator.PageTransportTimeoutException: Encountered too many errors talking to a worker node. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (http://10.16.130.19:8888/v1/task/20150805_115215_00013_eha94.2.6/results/20150805_115215_00013_eha94.1.6/0 - requests failed for 60.01s)
  at com.facebook.presto.operator.HttpPageBufferClient$1.onFailure(HttpPageBufferClient.java:325)
  at com.google.common.util.concurrent.Futures$6.run(Futures.java:1310)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.util.concurrent.TimeoutException
  at com.google.common.base.Throwables.propagate(Throwables.java:160)
  at io.airlift.http.client.ResponseHandlerUtils.propagate(ResponseHandlerUtils.java:22)
  at com.facebook.presto.operator.HttpPageBufferClient$PageResponseHandler.handleException(HttpPageBufferClient.java:434)
  at com.facebook.presto.operator.HttpPageBufferClient$PageResponseHandler.handleException(HttpPageBufferClient.java:421)
  at io.airlift.http.client.jetty.JettyHttpClient$JettyResponseFuture.failed(JettyHttpClient.java:793)
  at io.airlift.http.client.jetty.JettyHttpClient$BufferingResponseListener.onComplete(JettyHttpClient.java:1027)
  at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:193)
  at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:185)
  at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:446)
  at org.eclipse.jetty.client.HttpReceiver.abort(HttpReceiver.java:517)
  at org.eclipse.jetty.client.HttpChannel.abortResponse(HttpChannel.java:130)
  at org.eclipse.jetty.client.HttpChannel.abort(HttpChannel.java:123)
  at org.eclipse.jetty.client.HttpExchange.abort(HttpExchange.java:259)
  at org.eclipse.jetty.client.HttpConversation.abort(HttpConversation.java:141)
  at org.eclipse.jetty.client.HttpRequest.abort(HttpRequest.java:704)
  at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.abort(HttpConnectionOverHTTP.java:157)
  at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.close(HttpConnectionOverHTTP.java:145)
  at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onReadTimeout(HttpConnectionOverHTTP.java:98)
  at org.eclipse.jetty.io.AbstractConnection.onFillInterestedFailed(AbstractConnection.java:173)
  at org.eclipse.jetty.io.AbstractConnection$ReadCallback$1.run(AbstractConnection.java:582)
  at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
  at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
  ... 1 more
Caused by: java.util.concurrent.TimeoutException
  ... 6 more
```

- Two nodes are not visible in discovery service for some time after this failure:
```
# before execution of inserts
presto:default> select * from system.runtime.nodes;
               node_id                |         http_uri         |         node_version          | coordinator | active 
--------------------------------------+--------------------------+-------------------------------+-------------+--------
 31f4540a-8ba8-4df9-9d34-60876a907ef9 | http://10.16.130.16:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 becb09f8-b229-4312-ae52-1a5969c7b860 | http://10.16.130.20:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 9042f9a7-197f-4b0c-bd0b-c5d0f9d22809 | http://10.25.22.149:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 e08d8bfa-d68b-414e-8c72-1b3e1f0a5c7c | http://10.16.130.2:8888  | presto-main:0.112-64-g99fc9d2 | true        | true   
 ea82783a-5314-4dd3-b2e4-6a87732f5acd | http://10.16.130.19:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 01832273-8d15-4964-a281-163aed92a6a6 | http://10.16.130.3:8888  | presto-main:0.112-64-g99fc9d2 | false       | true   
 a029f63a-0cb7-4e91-a2c2-026ff28d1f54 | http://10.16.130.15:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 25d11ad3-aae9-4985-b10d-72938353da39 | http://10.16.130.18:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
(8 rows)

Query 20150805_115118_00008_eha94, FINISHED, 2 nodes
Splits: 2 total, 2 done (100.00%)
0:00 [8 rows, 726B] [19 rows/s, 1.7KB/s]

# after execution of inserts - later on they show up again
presto:default> select * from system.runtime.nodes;
               node_id                |         http_uri         |         node_version          | coordinator | active 
--------------------------------------+--------------------------+-------------------------------+-------------+--------
 31f4540a-8ba8-4df9-9d34-60876a907ef9 | http://10.16.130.16:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 becb09f8-b229-4312-ae52-1a5969c7b860 | http://10.16.130.20:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 e08d8bfa-d68b-414e-8c72-1b3e1f0a5c7c | http://10.16.130.2:8888  | presto-main:0.112-64-g99fc9d2 | true        | true   
 01832273-8d15-4964-a281-163aed92a6a6 | http://10.16.130.3:8888  | presto-main:0.112-64-g99fc9d2 | false       | true   
 a029f63a-0cb7-4e91-a2c2-026ff28d1f54 | http://10.16.130.15:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
 25d11ad3-aae9-4985-b10d-72938353da39 | http://10.16.130.18:8888 | presto-main:0.112-64-g99fc9d2 | false       | true   
(6 rows)
```

- I can see in logs quite many exceptions: `java.io.IOException: Broken pipe` and `java.util.concurrent.TimeoutException` in airlift http client. During execution of those queries presto becomes unresponsive. I have dumped thread from one of the workers when this happens: https://gist.github.com/sogorkis/994acd7205cb2003a414

- Very strange thing is that parallel inserts are working well for `text` input format:
```
INSERT INTO hive.default.lineitem_insert_test_1 SELECT * FROM "hive"."tpch_100gb_text"."lineitem" WHERE quantity < 6;
INSERT INTO hive.default.lineitem_insert_test_2 SELECT * FROM "hive"."tpch_100gb_text"."lineitem" WHERE quantity < 6;
```

dain commented:
@sogorkis PAGE_TRANSPORT_TIMEOUT almost always mean the machines are having very long GC pauses (or they crashed).  Writing Hive file formats generates a huge amount of garbage which puts a lot of pressure on the GC. Normally, G1 can keep up with this, but if you have too many worker threads you can overwhelm G1.  I suggest tuning on the GC logs to see what is going on.  Also, you can reduce `task.max-worker-threads`.  In our Raptor system, we also externally limit the number of concurrent writer jobs.

This is an area we are actively working on and this PR represents the first step.  I expect that we will need to write a new ORC writer which is as performant as the rest of Presto.

sogorkis commented:
You were right, there seem to be long full GC happening:
```
2015-08-06T08:36:21.537-0400: 118.268: [Full GC (Allocation Failure)  14G->13G(16G), 22.4438964 secs]
   [Eden: 0.0B(816.0M)->0.0B(816.0M) Survivors: 0.0B->0.0B Heap: 14.7G(16.0G)->13.3G(16.0G)], [Metaspace: 69794K->69790K(1116160K)]
 [Times: user=38.79 sys=0.16, real=22.45 secs] 
2015-08-06T08:36:43.981-0400: 140.712: Total time for which application threads were stopped: 22.4446957 seconds, Stopping threads took: 0.0000997 seconds
```

yuananf commented:
Can this PR control the size of each ORC file written?
Also need a syntax like hive's `insert overwrite`.

KBP-TDC commented:
Not sure about "insert overwrite", it is not common in other SQL engines and also will make it difficult to manage table privileges once we get proper security (typically you want to separate permissions for INSERT and DELETE operations).

Having something like TRUNCATE to quickly empty an existing table seems desirable (e.g., http://www.postgresql.org/docs/9.4/static/sql-truncate.html)

dain commented:
@yuananf The size of orc files written is controlled by the standard hadoop configuration options (we haven't changed anything in the hadoop writers).

We have no plans to add `INSERT OVERWRITE`, but we are working on adding `DELETE`.  The first version of `DELETE` will only work when the `WHERE` clause can be exactly matched to the table partitioning.

@KBP-TDC, I agree we should add `TRUNCATE`.  If there isn't an issue for that, can you add one?

KBP-TDC commented:
Done: https://github.com/facebook/presto/issues/3424

haozhun commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/InMemoryHiveMetastore.java:
```diff
@@ -287,6 +377,58 @@ private static boolean isParentDir(File directory, File baseDirectory)
+        {
+            this.schemaName = schemaName.toLowerCase(US);
+            this.tableName = tableName.toLowerCase(US);
+            this.partitionName = partitionName.toLowerCase(US);
```
Remove `toLowerCase`

electrum commented:
Looks good

dain commented:
@electrum I added three small commits to the middle to address the bigger comments:
- Rewrite create table and insert rollback to be more safe 
- Improve insert tests
- Add Hive rollback tests

Can you review these? I'll squash them into the insert commit once reviewed.

haozhun commented:
@dain InMemoryHiveMetastore line 392:
```
-            this.partitionName = partitionName.toLowerCase(US);
+            this.partitionName = partitionName;
```

haozhun commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -1440,94 +1513,251 @@ protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storag
+        for (int i = 0; i < 3; i++) {
+            // begin the insert
+            ConnectorTableHandle tableHandle = getTableHandle(tableName);
+            ConnectorInsertTableHandle insertTableHandle = metadata.beginInsert(SESSION, tableHandle);
```
Use `insertData` instead?

haozhun commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -1440,94 +1513,251 @@ protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storag
+            MaterializedResult result = readTable(tableHandle, columnHandles, SESSION, TupleDomain.all(), OptionalInt.empty(), Optional.of(storageFormat));
+            assertEqualsIgnoreOrder(result.getMaterializedRows(), resultBuilder.build().getMaterializedRows());
+
+            assertEqualsIgnoreOrder(result.getMaterializedRows(), resultBuilder.build().getMaterializedRows());
```
duplicate check

buremba commented:
:100:
